### ISSUE TITLE
Warning about electron's removal of Node.js vm module

### ISSUE THREAD

**Issue by GreedyWhale** [bytenode/bytenode#107]
Hi, I am using bytenode to encrypt my electron application source code and it works very well, here is the version I am currently using.

```
"bytenode": "^1.1.7"
"electron": "^12.0.0"
```

When I packaged the app, electron gave me a warning that

> The vm module of Node.js is deprecated in the renderer process and will be removed.

This is electron's [Release Notes Persisted](https://github.com/electron/electron/pull/26087).

I would like to ask if electron really removed the vm module of Node.js in the renderer process, will it be affected if I continue to use bytenode？

**Comment by OsamaAbbas** [comment]
Bytenode depends on `vm` in order to work. So removing it may break bytenode.

I can think of a workaround if `vm` is removed, but I need to try it first to make sure it works.

**Comment by aaxx255** [comment]
Hello, how did you use it. Is it possible to ask you some guidelines. It is best to have a detailed tutorial. thank you very much!

**Comment by dopry** [comment]
@OsamaAbbas What was your idea for the work around?

**Comment by gpetrov** [comment]
Any progress on this @OsamaAbbas ?

More concerns about the vm module removal are mentioned here: https://github.com/electron/electron/issues/31724#issue-1045555972

Maybe you can join in - hope the Electron team change their mind

**Comment by OsamaAbbas** [comment]
> @OsamaAbbas What was your idea for the work around?

I tried to remember it but could not!

> Maybe you can join in - hope the Electron team change their mind

They won't, even if 100% of their users said otherwise. The same issue happened before multiple times where the team decides something and sticks to it no matter what everyone else wants or thinks.

**Comment by OsamaAbbas** [comment]
The only solution seems a bit radical, which is building electron from scratch and re-enabling the `vm` module.

**Comment by gpetrov** [comment]
How about using the V8 snapshot to load all compiled modules directly on load.

It seems possible and electron's mksnapshot and electron-link are using it already extensive.

**Comment by OsamaAbbas** [comment]
I have no experience with mksnapshot, but I will look into it. Thanks for the suggestion.

The problem is that I'm busy especially at the end of this year and the start of the new year. So I don't know where I will be able to chase this issue.

**Comment by gpetrov** [comment]
Here are some useful links about v8 snapshots - a lot of possibilities in there:

https://v8.dev/blog/custom-startup-snapshots
https://github.com/atom/electron-link/issues/1#issue-222957513
https://flight-manual.atom.io/behind-atom/sections/how-atom-uses-chromium-snapshots/
http://peterforgacs.github.io/2018/09/12/How-to-create-a-V8-snapshot-of-your-javascript-file/
https://github.com/inkdropapp/electron-v8snapshots-example

But that all can wait - for now enjoy the holidays @OsamaAbbas - your work is greatly appreciated!

**Comment by WHK102** [comment]
bytenode will stop having support for electron?

**Comment by OsamaAbbas** [comment]
> > > I can think of a workaround if vm is removed, but I need to try it first to make sure it works.
> >
> > @OsamaAbbas What was your idea for the work around?
> 
> I tried to remember it but could not!

Almost a year ago I was not able to remember what the workaround that I had to restore bytenode functionality if electron goes with removing vm module.

Strangely enough, I remembered it suddenly yesterday, so I wanted to record it here before I forget it again!

Depending on how electron would remove the support for vm in the renderer process, we have two options:

1. Just replace require('vm'); with require('node:vm'). This trivial solution works for now and it does not trigger the deprecation warning. ( This is because how they are intercepting the require request, they only look for vm specifically. See it here [reset-search-paths.ts#L10](https://github.com/electron/electron/blob/7493062555b9eb4e6cc47d628a4831c121192f1f/lib/common/reset-search-paths.ts#L10) ).

2. If electron decides to take a more radical approach by blocking every possible way to reach to the vm module (which seems unreasonable and hopefully unlikely), the other option will be creating a custom `vm` module. It's a hacky solution and I have not try this workaround (fully) yet, but it should work.

___

> bytenode will stop having support for electron?

Hopefully not.
